### PR TITLE
[SYCL][CUDA][libclc] Added atomics with scopes and memory orders

### PR DESCRIPTION
**Pull request by t4c1**
Added libclc implementations for CUDA atomics, including for various scopes and memory orders. They are implemented using LLVM intrinsics and exposed as clang builtins, which are than used to implement functions in libclc.

Sorry for a huge pull request, but I think for these changes it make sense they are reviewed together.

EDIT: llvm-test-suite PR with tests for this is https://github.com/intel/llvm-test-suite/pull/534

### CODE REVIEW

**Review by Naghasan — DISMISSED** [review]
Looks way better. Thanks.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGBuiltin.cpp:17709]: Looks like this code is duplicated several times. Can this be factored out similar to `MakeScopedAtomic `

File: libclc/ptx-nvidiacl/libspirv/atomic/atomic_helpers.h
Comment (steffenlarsen) [inline libclc/ptx-nvidiacl/libspirv/atomic/atomic_helpers.h:15]: This looks unused.

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (steffenlarsen) [inline clang/include/clang/Basic/BuiltinsNVPTX.def:667]: In the PTX instruction the scope appears before the semantic. I am okay with the space appearing at the end, but I think it would make more sense to have the other other options occur in the same order as the instruction.

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (steffenlarsen) [inline clang/include/clang/Basic/BuiltinsNVPTX.def:1038]: I cannot help but wonder if the builtins could select the right intrinsics without the space explicitly stated in the builtin's name, i.e. inferring it from the pointer's address space. Or maybe even as late as instruction selection?
@Naghasan - Do you have insight into whether or not that is possible?

**Review by steffenlarsen — COMMENTED** [review]
A few comments, but I think it's good overall. That said, I agree with @bader that everything in clang/* and llvm/* should be reviewed and merged upstream (see https://llvm.org/docs/Contributing.html).

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (t4c1) [reply clang/include/clang/Basic/BuiltinsNVPTX.def:667]: > In the PTX instruction the scope appears before the semantic.

Do you mean the other way around? In builtins in this PR the scope appears before the semantic and in PTX ISA the semantic appears before the scope.

I would not change the order of anything but the semantics, as existing builtins have other parts in that order. There are no builtins for atomics with semantic, so I guess we can put that in any place.

Comment (steffenlarsen) [reply clang/include/clang/Basic/BuiltinsNVPTX.def:667]: > Do you mean the other way around? In builtins in this PR the scope appears before the semantic and in PTX ISA the semantic appears before the scope.

Sorry, that is what I meant yes!

> I would not change the order of anything but the semantics, as existing builtins have other parts in that order. There are no builtins for atomics with semantic, so I guess we can put that in any place.

I agree, if other builtins use a specific order I think it should be kept as consistent as possible, but as you mention there are no other other builtins with atomic semantics (at least that I know off) so personally I would prefer to have them as similar to the corresponding PTX instruction as possible to make the mapping simpler.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (t4c1) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Yes, done.

File: libclc/ptx-nvidiacl/libspirv/atomic/atomic_helpers.h
Comment (t4c1) [reply libclc/ptx-nvidiacl/libspirv/atomic/atomic_helpers.h]: Removed.

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (t4c1) [reply clang/include/clang/Basic/BuiltinsNVPTX.def]: I refactored this as requested.

Comment (steffenlarsen) [reply clang/include/clang/Basic/BuiltinsNVPTX.def]: Thanks!

**Comment by t4c1** [comment]
> That said, I agree with @bader that everything in clang/* and llvm/* should be reviewed and merged upstream (see https://llvm.org/docs/Contributing.html).

I requested a review, but forgot to post a link to it. It is here: https://reviews.llvm.org/D112718

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (elizabethandrews) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Thanks!

**Review by elizabethandrews — APPROVED** [review]
I'm not familiar with this, so I cannot comment on the functionality. FE changes LGTM otherwise. Please wait for review from someone more familiar with this before merge. If this is merged to llvm-project instead, please apply review comments to patch uploaded there.

**Review by smanna12 — APPROVED** [review]
FE Changes look good to me.

**Comment by bader** [comment]
@t4c1, it looks like the build is broken. Could you take a look at the failures here: http://ci.llvm.intel.com:8010/#builders/37/builds/14729, please?

**Comment by t4c1** [comment]
Thank you for the reminder. I forgot to make changes in libclc after modifying the builtins. Should be fixed now.

**Review by bader — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Pull request merged**
